### PR TITLE
boards/nrf52840dk: split into lib+bin crates for configuration/tutorial boards

### PR DESCRIPTION
**Pull request by lschuermann**
### Pull Request Overview

This pull request separates the `nrf52840dk` crate into a `lib` and `bin` crate, which allows us to depend on it from within other board crates. This enables de-duplicating the base board definition logic between in tutorial- and/or configuration-board crates.

### Testing Strategy

This pull request was tested by compiling and using it in a tutorial board definition.

### TODO or Help Wanted

Because we rename the `main.rs` to `lib.rs` and create a new `main.rs`, git doesn't really track this as a rename. As such, we should avoid introducing conflicts and re-do this porting if a conflicting PR lands.

### Documentation Updated

- [X] Updated the relevant files in `/docs`, or no updates are required.

### Formatting

- [X] Ran `make prepush`.

### CODE REVIEW

**Comment by lschuermann** [comment]
On @bradjc's comment on #3979:

> While I wish we didn't have to move code to a new file, this seems like a necessary step forward. I think it is worth it to have a convention for supporting boards with shields/peripherals.

While I believe this is the best solution available to us, I'd propose to treat this as an experiment. For instance, I've already encountered some unexpected behavior: process loading finalizes grants, so instantiating a new driver after it would panic. Hence we move the process loading into the bin crate, and duplicate it for the configuration boards.

In retrospect this particular change makes sense, as it's quite orthogonal to base-peripheral instantiation and it's well-possible we'll want to use a different loading strategy for configuration boards (e.g., with app signing). But we may encounter more such cases, and I'm totally open to reverting this if this is too inconvenient in the end.

**Review by bradjc — APPROVED** [review]
I added a doc to describe this design and provide as least some insight into why this board is different from the rest.

**Comment by bradjc** [comment]
Anyone else able to look at this?

**Comment by ppannuto** [comment]
For others looking and not wanting to recall the somewhat arcane git syntax for tracking diffs across a rename, here's the changes between the old main.rs and new lib.rs:

```diff
diff --git a/boards/nordic/nrf52840dk/src/main.rs b/boards/nordic/nrf52840dk/src/lib.rs
index 8590e887f..13fcdcc0f 100644
--- a/boards/nordic/nrf52840dk/src/main.rs
+++ b/boards/nordic/nrf52840dk/src/lib.rs
@@ -68,12 +68,9 @@
 //! | P0.27 | P24 17 | I2C SCL  |
 
 #![no_std]
-// Disable this attribute when documenting, as a workaround for
-// https://github.com/rust-lang/rust/issues/62184.
-#![cfg_attr(not(doc), no_main)]
 #![deny(missing_docs)]
 
-use core::ptr::{addr_of, addr_of_mut};
+use core::ptr::addr_of;
 
 use capsules_core::virtualizers::virtual_alarm::VirtualMuxAlarm;
 use capsules_extra::net::ieee802154::MacAddress;
@@ -87,6 +84,7 @@ use kernel::platform::{KernelResources, SyscallDriverLookup};
 use kernel::scheduler::round_robin::RoundRobinSched;
 #[allow(unused_imports)]
 use kernel::{capabilities, create_capability, debug, debug_gpio, debug_verbose, static_init};
+use nrf52840::chip::Nrf52DefaultPeripherals;
 use nrf52840::gpio::Pin;
 use nrf52840::interrupt_service::Nrf52840DefaultPeripherals;
 use nrf52_components::{UartChannel, UartPins};
@@ -130,6 +128,7 @@ const DEFAULT_CTX_PREFIX_LEN: u8 = 8; //Length of context for 6LoWPAN compressio
 const DEFAULT_CTX_PREFIX: [u8; 16] = [0x0_u8; 16]; //Context for 6LoWPAN Compression
 
 /// Debug Writer
+#[allow(unused)] // TODO: this throws unused warnings in `make ci-job-capsules`
 pub mod io;
 
 // Whether to use UART debugging or Segger RTT (USB) debugging.
@@ -137,14 +136,11 @@ pub mod io;
 // - Set to true to use Segger RTT over USB.
 const USB_DEBUGGING: bool = false;
 
-// State for loading and holding applications.
-// How should the kernel respond when a process faults.
-const FAULT_RESPONSE: kernel::process::PanicFaultPolicy = kernel::process::PanicFaultPolicy {};
+/// Number of concurrent processes this platform supports.
+pub const NUM_PROCS: usize = 8;
 
-// Number of concurrent processes this platform supports.
-const NUM_PROCS: usize = 8;
-
-static mut PROCESSES: [Option<&'static dyn kernel::process::Process>; NUM_PROCS] =
+/// Process array of this platform.
+pub static mut PROCESSES: [Option<&'static dyn kernel::process::Process>; NUM_PROCS] =
     [None; NUM_PROCS];
 
 static mut CHIP: Option<&'static nrf52840::chip::NRF52<Nrf52840DefaultPeripherals>> = None;
@@ -224,7 +220,8 @@ pub struct Platform {
     rng: &'static RngDriver,
     adc: &'static capsules_core::adc::AdcDedicated<'static, nrf52840::adc::Adc<'static>>,
     temp: &'static TemperatureDriver,
-    ipc: kernel::ipc::IPC<{ NUM_PROCS as u8 }>,
+    /// The IPC driver.
+    pub ipc: kernel::ipc::IPC<{ NUM_PROCS as u8 }>,
     analog_comparator: &'static capsules_extra::analog_comparator::AnalogComparator<
         'static,
         nrf52840::acomp::Comparator<'static>,
@@ -322,6 +319,7 @@ pub unsafe fn start() -> (
     &'static kernel::Kernel,
     Platform,
     &'static nrf52840::chip::NRF52<'static, Nrf52840DefaultPeripherals<'static>>,
+    &'static Nrf52DefaultPeripherals<'static>,
 ) {
     //--------------------------------------------------------------------------
     // INITIAL SETUP
@@ -400,8 +398,6 @@ pub unsafe fn start() -> (
 
     // Create capabilities that the board needs to call certain protected kernel
     // functions.
-    let process_management_capability =
-        create_capability!(capabilities::ProcessManagementCapability);
     let memory_allocation_capability = create_capability!(capabilities::MemoryAllocationCapability);
     let gpio_port = &nrf52840_peripherals.gpio_port;
 
@@ -930,46 +926,5 @@ pub unsafe fn start() -> (
     debug!("Initialization complete. Entering main loop\r");
     debug!("{}", &*addr_of!(nrf52840::ficr::FICR_INSTANCE));
 
-    // These symbols are defined in the linker script.
-    extern "C" {
-        /// Beginning of the ROM region containing app images.
-        static _sapps: u8;
-        /// End of the ROM region containing app images.
-        static _eapps: u8;
-        /// Beginning of the RAM region for app memory.
-        static mut _sappmem: u8;
-        /// End of the RAM region for app memory.
-        static _eappmem: u8;
-    }
-
-    kernel::process::load_processes(
-        board_kernel,
-        chip,
-        core::slice::from_raw_parts(
-            core::ptr::addr_of!(_sapps),
-            core::ptr::addr_of!(_eapps) as usize - core::ptr::addr_of!(_sapps) as usize,
-        ),
-        core::slice::from_raw_parts_mut(
-            core::ptr::addr_of_mut!(_sappmem),
-            core::ptr::addr_of!(_eappmem) as usize - core::ptr::addr_of!(_sappmem) as usize,
-        ),
-        &mut *addr_of_mut!(PROCESSES),
-        &FAULT_RESPONSE,
-        &process_management_capability,
-    )
-    .unwrap_or_else(|err| {
-        debug!("Error loading processes!");
-        debug!("{:?}", err);
-    });
-
-    (board_kernel, platform, chip)
-}
-
-/// Main function called after RAM initialized.
-#[no_mangle]
-pub unsafe fn main() {
-    let main_loop_capability = create_capability!(capabilities::MainLoopCapability);
-
-    let (board_kernel, platform, chip) = start();
-    board_kernel.kernel_loop(&platform, chip, Some(&platform.ipc), &main_loop_capability);
+    (board_kernel, platform, chip, base_peripherals)
 }
```

File: boards/nordic/nrf52840dk/src/lib.rs
Comment (ppannuto) [inline boards/nordic/nrf52840dk/src/lib.rs:132]: Not necessarily going to block this PR for this — but is this something that could be fixed?

The debug infrastructure is a bit fragile with the off-common-path statics, and I'm a bit worried this is hiding a real issue.

Comment (lschuermann) [reply boards/nordic/nrf52840dk/src/lib.rs]: Fixed. I had a hard time tracking down the root cause of this issue; it turned out to be conditional compilation, of course. `make ci-job-capsules` apparently passes `cfg(test)`, which does not select our custom panic handler. The reason why this didn't trip CI before is that `ci-job-capsules` reall]y tests _every_ crate (beyond just capsules), but only _library_ ones. With this crate being promoted to a library crate, it was now being "tested" and hence raised unused warnings.

**Review by bradjc — APPROVED** [review]

**Comment by bradjc** [comment]
Just noting that I do not think we should use this to re-use top-level boards for configuration boards. Maintainers of upstream boards should not have to worry if changes will affect testing boards.

**Pull request merged**
